src/components/lib/shm_bm/shm_bm.h

```diff
@@ -0,0 +1,271 @@
+#define SHM_BM_ALIGN (1 << 22)
+#define SHM_BM_BITS_TO_WORDS(bits) (round_up_to_pow2(bits, SHM_BM_BITMAP_BLOCK) / SHM_BM_BITMAP_BLOCK)
+#define SHM_BM_SERVICE_TABLE_SIZE MAX_NUM_COMPS
+
+typedef void *        shm_bm_t;
+typedef int           shm_reqtok_t;
+typedef unsigned int  shm_objid_t;
+
+struct shm_service_tbl_ent {
+	invtoken_t authtok;
+	void *     shmptr;
+};
+
+typedef unsigned long word_t;
+typedef unsigned char refcnt_t;
+typedef unsigned char byte_t;
+
+#define SHM_BM_REFC(shm, nobj) ((refcnt_t *) ((byte_t *) shm + SHM_BM_BITS_TO_WORDS(nobj) * sizeof (word_t)))
+#define SHM_BM_DATA(shm, nobj) ((byte_t *)  (SHM_BM_REFC(shm, nobj) + (unsigned int) nobj))
+
+static inline void
+__shm_bm_set_contig(word_t *bm, int offset)
+{
+	int i, n, ind;
+
+	ind = offset / SHM_BM_BITMAP_BLOCK;
+	offset %= SHM_BM_BITMAP_BLOCK;
+
+	for (i = 0; i < ind; i++) {
+		bm[i] = ~0x0ul;
+	}
+
+	n = SHM_BM_BITMAP_BLOCK - offset;
+	bm[ind] = ~((1ul << n) - 1); // set most sig n bits of bm[ind]
+}
+
+/* Find the first nonzero word in the inputted bitmap */
+static inline int
+__shm_bm_next_free_word(word_t *bm, unsigned long nwords)
+{
+	/* the index of the word in the bitmap where the first set bit is found */
+	int idx;
+
+	for (idx = 0; idx < (int) nwords; idx++) {
+		if (*bm != 0) return idx;
+		bm++;
+	}
+
+	/* all bits set to zero */
+	return -1;
+}
+
+static inline cbuf_t 
+__shm_bm_clt_create(shm_bm_t *shm, size_t objsz, unsigned int nobj)
+{
+	size_t                bitmap_sz, refcnt_sz, data_sz;
+	size_t                alloc;
+	cbuf_t                id;
+
+	if (nobj <= 0) return 0;
+
+	bitmap_sz = SHM_BM_BITS_TO_WORDS(nobj) * sizeof (word_t);
+	refcnt_sz = nobj;
+	data_sz   = nobj * objsz;
+
+	alloc = bitmap_sz + refcnt_sz + data_sz;
+
+	id  = memmgr_shared_page_allocn_aligned(round_up_to_page(alloc)/PAGE_SIZE, SHM_BM_ALIGN, (vaddr_t *) shm);
+	if (id == 0) return 0;
+	memset(*shm, 0, round_up_to_page(alloc));
+
+	// set nobj bits to free
+	__shm_bm_set_contig((word_t *) *shm, nobj);
+	return id;
+}
+
+static inline void * 
+__shm_bm_clt_alloc(shm_bm_t shm, shm_objid_t *objid, size_t objsz, unsigned int nobj)
+{
+	int     freebit, idx , offset, lz;
+	word_t  word; 
+	word_t *bm;
+
+	/* 
+	 * Find a free space; could be preempted. Find the next word in
+	 * the bitmap with a free bit and get its index in the bitmap, and 
+	 * the offset of the free bit in that word. Then try to atomically 
+	 * xchg the bitmap with the freebit set to 0
+	 */
+	bm = (word_t *) shm;
+	do {
+		idx = __shm_bm_next_free_word(bm, SHM_BM_BITS_TO_WORDS(nobj));
+		if (idx == -1) return 0;
+		word   = bm[idx];
+		lz     = __builtin_clzl(word); 
+		offset = SHM_BM_BITMAP_BLOCK - lz - 1;
+	} while (!cos_cas(bm + idx, word, word & ~(1ul << offset)));
+
+	freebit = lz + (idx * SHM_BM_BITMAP_BLOCK);
+
+	cos_faab(SHM_BM_REFC(shm, nobj) + freebit, 1);
+
+	*objid = (shm_objid_t) freebit;
+	return SHM_BM_DATA(shm, nobj) + (freebit * objsz);
+}
+
+static inline shm_reqtok_t
```

Comment:
These functions around the mapping/authentication are likely another layer of abstraction. As they might be implemented using different techniques, we'll want to pull them out at some point. However, for now, I think this is OK here.

Reply:
If you could, add a TODO comment that says to move these into a separate abstraction that could allow allocation with e.g.. ps as well.